File: C++/minimize-result-by-adding-parentheses-to-expression.cpp
// Time:  O(n^2)
// Space: O(1)

// brute force
class Solution {
public:
    string minimizeResult(string expression) {
        const auto& stoi = [](const auto& s, int i, int j) {
            if (i == j) {
                return 1;
            }
            int result = 0;
            for (int k = i; k < j; ++k) {
                result = result * 10 + (s[k] - '0');
            }
            return result;
        };
        const auto& pos = expression.find('+');
        int left = stoi(expression, 0, pos), right = stoi(expression, pos + 1, size(expression));
        int min_val = numeric_limits<int>::max();
        pair<int, int> best;
        const int base2_init = pow(10, size(expression) - (pos + 1) - 1);
        for (int i = 0, base1 = pow(10, pos); i < pos; ++i, base1 /= 10) {
            for (int j = pos + 1, base2 = base2_init; j < size(expression); ++j, base2 /= 10) {
                const int a = left / base1, b = left % base1;
                const int c = right / base2, d = right % base2;
                const auto& val = max(a, 1) * (b + c) * max(d, 1);
                if (val < min_val) {
                    min_val = val;
                    best = {i, j};
                }
            }
        }
        string result;
        for (int i = 0; i < size(expression); ++i) {
            if (i == best.first) {
                result.push_back('(');
            }
            result.push_back(expression[i]);
            if (i == best.second) {
                result.push_back(')');
            }
        }
        return result;
    }
};

// Time:  O(n^2)
// Space: O(n)
// brute force
class Solution2 {
public:
    string minimizeResult(string expression) {
        const auto& pos = expression.find('+');
        int left = stoi(expression.substr(0, pos)), right = stoi(expression.substr(pos + 1));  // Space: O(n)
        int min_val = numeric_limits<int>::max();
        pair<int, int> best;
        const int base2_init = pow(10, size(expression) - (pos + 1) - 1);
        for (int i = 0, base1 = pow(10, pos); i < pos; ++i, base1 /= 10) {
            for (int j = pos + 1, base2 = base2_init; j < size(expression); ++j, base2 /= 10) {
                const int a = left / base1, b = left % base1;
                const int c = right / base2, d = right % base2;
                const auto& val = max(a, 1) * (b + c) * max(d, 1);
                if (val < min_val) {
                    min_val = val;
                    best = {i, j};
                }
            }
        }
        string result = expression.substr(0, best.first);  // Space: O(n)
        result += "(";
        result += expression.substr(best.first, best.second + 1 - best.first);  // Space: O(n)
        result += ")";
        result += expression.substr(best.second + 1);  // Space: O(n)
        return result;
    }
};

// Time:  O(n^3)
// Space: O(n)
// brute force
class Solution3 {
public:
    string minimizeResult(string expression) {
        const auto& pos = expression.find('+');
        int min_val = numeric_limits<int>::max();
        pair<int, int> best;
        for (int i = 0; i < pos; ++i) {
            for (int j = pos + 1; j < size(expression); ++j) {
                const auto& val = (i != 0 ? stoi(expression.substr(0, i)) : 1) *
                                  (stoi(expression.substr(i, pos - i)) + stoi(expression.substr(pos + 1, j + 1 - (pos + 1)))) *
                                  (j + 1 != size(expression) ? stoi(expression.substr(j + 1)) : 1);  // Space: O(n)
                if (val < min_val) {
                    min_val = val;
                    best = {i, j};
                }
            }
        }
        string result = expression.substr(0, best.first);  // Space: O(n)
        result += "(";
        result += expression.substr(best.first, best.second + 1 - best.first);  // Space: O(n)
        result += ")";
        result += expression.substr(best.second + 1);  // Space: O(n)
        return result;
    }
};

File: Python/minimize-result-by-adding-parentheses-to-expression.py
# Time:  O(n^2)
# Space: O(1)

import itertools


# brute force
class Solution(object):
    def minimizeResult(self, expression):
        """
        :type expression: str
        :rtype: str
        """
        def stoi(s, i, j):
            result = 0
            for k in xrange(i, j):
                result = result*10+(ord(s[k])-ord('0'))
            return result

        best = None
        min_val = float("inf")
        pos = expression.index('+')
        left, right = stoi(expression, 0, pos), stoi(expression, pos+1, len(expression))
        base1, base2_init = 10**pos, 10**(len(expression)-(pos+1)-1)
        for i in xrange(pos):
            base2 = base2_init
            for j in xrange(pos+1, len(expression)):
                a, b = divmod(left, base1)
                c, d = divmod(right, base2)
                val = max(a, 1)*(b+c)*max(d, 1)
                if val < min_val:
                    min_val = val
                    best = (i, j)
                base2 //= 10
            base1 //= 10
        return "".join(itertools.chain((expression[i] for i in xrange(best[0])),
                                       '(', (expression[i] for i in xrange(best[0], best[1]+1)), ')',
                                       (expression[i] for i in xrange(best[1]+1, len(expression)))))


# Time:  O(n^2)
# Space: O(n)
# brute force
class Solution2(object):
    def minimizeResult(self, expression):
        """
        :type expression: str
        :rtype: str
        """
        best = None
        min_val = float("inf")
        pos = expression.index('+')
        left, right = int(expression[0:pos]), int(expression[pos+1:])  # Space: O(n)
        base1, base2_init = 10**pos, 10**(len(expression)-(pos+1)-1)
        for i in xrange(pos):
            base2 = base2_init
            for j in xrange(pos+1, len(expression)):
                a, b = divmod(left, base1)
                c, d = divmod(right, base2)
                val = max(a, 1)*(b+c)*max(d, 1)
                if val < min_val:
                    min_val = val
                    best = (i, j)
                base2 //= 10
            base1 //= 10
        return "".join([expression[:best[0]], '(', expression[best[0]:best[1]+1], ')', expression[best[1]+1:]])  # Space: O(n)

    
# Time:  O(n^3)
# Space: O(n)
# brute force
class Solution3(object):
    def minimizeResult(self, expression):
        """
        :type expression: str
        :rtype: str
        """
        best = None
        min_val = float("inf")
        pos = expression.index('+')
        for i in xrange(pos):
            for j in xrange(pos+1, len(expression)):
                val = (int(expression[:i] or "1")*
                       (int(expression[i:pos])+int(expression[pos+1:j+1]))*
                       int(expression[j+1:] or "1"))  # Space: O(n)
                if val < min_val:
                    min_val = val
                    best = (i, j)
        return "".join([expression[:best[0]], '(', expression[best[0]:best[1]+1], ')', expression[best[1]+1:]])  # Space: O(n)
